include/cetl/any.hpp

```diff
@@ -26,69 +26,226 @@ class any;
     // Also, we need to ensure that the buffer is at least 1 byte long.
     alignas(Alignment) char buffer_[std::max(Footprint, 1UL)];
-};
 
-// Movable case.
-template <std::size_t Footprint, bool Copyable, std::size_t Alignment>
-struct base_move<Footprint, Copyable, true, Alignment> : base_copy<Footprint, Copyable, Alignment>
-{
-public:
-    constexpr base_move()                      = default;
-    base_move(base_move&&) noexcept            = default;
-    base_move& operator=(base_move&&) noexcept = default;
-};
+    // Holds type-erased value destroyer. `nullptr` if storage has no value stored.
+    void (*value_destroyer_)(void* self) = nullptr;
 
-// Non-movable case.
-template <std::size_t Footprint, bool Copyable, std::size_t Alignment>
-struct base_move<Footprint, Copyable, false, Alignment> : base_copy<Footprint, Copyable, Alignment>
-{
 public:
-    constexpr base_move()                      = default;
-    base_move(base_move&&) noexcept            = delete;
-    base_move& operator=(base_move&&) noexcept = delete;
-};
+    // Holds type-erased value copyer. `nullptr` when copy operation is not supported.
+    void (*value_copier_)(const void* src, void* dst) = nullptr;
 
-// Copyable case.
-template <std::size_t Footprint, std::size_t Alignment>
-struct base_copy<Footprint, true, Alignment> : base_storage<Footprint, Alignment>
-{
-public:
-    constexpr base_copy()                  = default;
-    base_copy(const base_copy&)            = default;
-    base_copy& operator=(const base_copy&) = default;
-};
+    // Holds type-erased value mover. `nullptr` when move operation is not supported.
+    void (*value_mover_)(void* src, void* dst) = nullptr;
+
+    base_storage() = default;
+
+    CETL_NODISCARD void* get_raw_storage() noexcept
+    {
+        return static_cast<void*>(buffer_);
+    }
+
+    CETL_NODISCARD const void* get_raw_storage() const noexcept
+    {
+        return static_cast<const void*>(buffer_);
+    }
+
+    CETL_NODISCARD bool has_value() const noexcept
+    {
+        return nullptr != value_destroyer_;
+    }
+
+    template <typename Tp>
+    void make_handlers() noexcept
+    {
+        assert(nullptr == value_destroyer_);
+
+        value_destroyer_ = [](void* const storage) {
+            const auto ptr = static_cast<Tp*>(storage);
+            ptr->~Tp();
+        };
+    }
+
+    template <typename ValueType>
+    CETL_NODISCARD void* get_ptr() noexcept
+    {
+        if (!has_value())
+        {
+            return nullptr;
+        }
+
+        // TODO: Add RTTI check here.
+        return get_raw_storage();
+    }
 
+    template <typename ValueType>
+    CETL_NODISCARD const void* get_ptr() const noexcept
+    {
+        if (!has_value())
+        {
+            return nullptr;
+        }
+
+        // TODO: Add RTTI check here.
+        return get_raw_storage();
+    }
+
+    void copy_from(const base_storage& src)
```

Comment:
If `this` already contains a value, what happens to it? The type-erased copy handler does not (cannot really) invoke the dtor.

Same question for the move case.

Reply:
There are no such paths - these `copy_from` & `move_from` are always called ONLY from empty `this`. I'm gonna add such `assert`-s (like `assert(!has_value());`) to this functions.

Reply:
Also, after addressing below comment...
namely: _> This will allow you to make copy_from and move_from private..._
both of these function are now private and called only 2 times (previously 5), so `this` is definitely empty at this point.

Reply:
Actually, I was wrong about `base_copy& operator=(const base_copy& other)` ! There is indeed memory leak in case `this` already contains a value. I'm extending tests with simple verification that number constructs equals to number of destructs in the end.